x509/cert_validations.go

```diff
@@ -46,28 +48,43 @@ func validateCertChain(certChain []*x509.Certificate, expectedLeafEku x509.ExtKe
 		if signingTime != nil && (signingTime.Before(cert.NotBefore) || signingTime.After(cert.NotAfter)) {
 			return fmt.Errorf("certificate with subject %q was not valid at signing time of %s", cert.Subject, signingTime.UTC())
 		}
-
 		if err := cert.CheckSignature(cert.SignatureAlgorithm, cert.RawTBSCertificate, cert.Signature); err != nil {
-			return err
+			return fmt.Errorf("self-signed signing certificate with subject %q failed at checking signature. Error: %v", cert.Subject, err)
+		}
+		if err := validateLeafCertificate(cert, expectedLeafEku); err != nil {
+			return fmt.Errorf("self-signed signing certificate is invalid. Error: %v", err)
 		}
-		return validateLeafCertificate(cert, expectedLeafEku)
+		return nil
 	}
 
 	for i, cert := range certChain {
 		if signingTime != nil && (signingTime.Before(cert.NotBefore) || signingTime.After(cert.NotAfter)) {
 			return fmt.Errorf("certificate with subject %q was not valid at signing time of %s", cert.Subject, signingTime.UTC())
 		}
-
+		selfSigned, selfSignedError := isSelfSigned(cert)
```

Comment:
Can this can be moved inside `if i == len(certChain)-1 {` ?

Reply:
But it is also used under the `else` statement: https://github.com/notaryproject/notation-core-go/blob/0cb6e7508c67ee9a3b3a2e680ae7841ec37fdd32/x509/cert_validations.go#L67